File: AgoraChat/Code/Controllers/Settings/ACDModifyAvatarViewController.m
//
//  ACDModifyAvatarViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/11/5.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDModifyAvatarViewController.h"
#import "ACDAvatarCollectionCell.h"

@interface ACDModifyAvatarViewController ()<UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout>

@property (nonatomic, strong) UICollectionView *collectionView;
@property (nonatomic, strong) UICollectionViewFlowLayout *collectionViewLayout;
@property (nonatomic, strong) NSMutableArray *itemArray;
@property (nonatomic, strong) NSIndexPath *selectedIndexPath;

@end

@implementation ACDModifyAvatarViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"done" style:UIBarButtonItemStylePlain target:self action:@selector(editAction)];
    
    [self placeAndLayoutSubviews];
    [self loadItems];
}


- (void)placeAndLayoutSubviews {
    [self.view addSubview:self.collectionView];
    [self.collectionView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
}

- (void)loadItems {
    for (NSInteger i = 1; i < 8; ++i) {
        NSString *imageName = [NSString stringWithFormat:@"defatult_avatar_%@",@(i)];
        [self.itemArray addObject:imageName];
    }
    
    [self.collectionView reloadData];
}

- (void)doneAction {
    if (self.selectedIndexPath && self.selectedBlock) {
        NSString *imageName = self.itemArray[self.selectedIndexPath.row];
        self.selectedBlock(imageName);
    }
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - UICollectionViewDataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.itemArray.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    ACDAvatarCollectionCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:[ACDAvatarCollectionCell reuseIdentifier] forIndexPath:indexPath];
    
    NSString *imageName = self.itemArray[indexPath.row];
    [cell.iconImageView setImage:ImageWithName(imageName)];
    
    if (self.selectedIndexPath.row == indexPath.row) {
        cell.selected = YES;
    }else {
        cell.selected = NO;
    }
    
    return cell;
}

#pragma mark - UICollectionViewDelegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    [collectionView deselectItemAtIndexPath:indexPath animated:YES];
    self.selectedIndexPath = indexPath;
    [self.collectionView reloadData];
}


#pragma mark getter and setter
- (UICollectionView *)collectionView {
    if (_collectionView == nil) {
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, KScreenHeight) collectionViewLayout:self.collectionViewLayout];
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
        
        [_collectionView registerClass:[ACDAvatarCollectionCell class] forCellWithReuseIdentifier:[ACDAvatarCollectionCell reuseIdentifier]];
    }
    return _collectionView;
}

- (UICollectionViewFlowLayout *)collectionViewLayout {
    UICollectionViewFlowLayout* flowLayout = [[UICollectionViewFlowLayout alloc] init];
    flowLayout.scrollDirection = UICollectionViewScrollDirectionVertical;
    CGFloat itemWidth = (KScreenWidth - 5.0 * 2)/2.0;
    flowLayout.itemSize = CGSizeMake(itemWidth, itemWidth);
    flowLayout.minimumLineSpacing = 5.0;
    flowLayout.minimumInteritemSpacing = 5.0;
    flowLayout.sectionInset = UIEdgeInsetsMake(0,
                                               0,
                                               0,
                                               0);
    return flowLayout;
}

- (NSMutableArray *)itemArray {
    if (_itemArray == nil) {
        _itemArray = NSMutableArray.new;
    }
    return _itemArray;
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Settings/ACDModifyAvatarViewController.swift
//
//  ACDModifyAvatarViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/13.
//

import UIKit

class ACDModifyAvatarViewController: UIViewController {

    private var collectionView: UICollectionView!
    private var itemArray: [String] = []
    private var selectedIndexPath: IndexPath?
    
    var selectedHandle: ((String) -> Void)?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Done".localized, style: .plain, target: self, action: #selector(doneAction))
        
        let itemWidth = (UIScreen.main.bounds.width - 10) / 2
        let layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .vertical
        layout.itemSize = CGSize(width: itemWidth, height: itemWidth)
        layout.minimumLineSpacing = 5
        layout.minimumInteritemSpacing = 5
        layout.sectionInset = UIEdgeInsets.zero
        
        self.collectionView = UICollectionView(frame: CGRect.zero, collectionViewLayout: layout)
        self.collectionView.dataSource = self
        self.collectionView.delegate = self
        self.collectionView.register(UINib(nibName: "ACDAvatarCollectionCell", bundle: nil), forCellWithReuseIdentifier: "cell")
        self.view.addSubview(self.collectionView)
        
        for i in 1..<8 {
            let imageName = "defatult_avatar_\(i)"
            self.itemArray.append(imageName)
        }
        self.collectionView.reloadData()
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        self.collectionView.frame = self.view.bounds
        if let layout = self.collectionView.collectionViewLayout as? UICollectionViewFlowLayout {
            let itemWidth = (self.view.bounds.width - 10) / 2
            layout.itemSize = CGSize(width: itemWidth, height: itemWidth)
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(false, animated: true)
    }
    
    @objc private func doneAction() {
        if let selectedIndexPath = self.selectedIndexPath, let selectedHandle = self.selectedHandle {
            let imageName = self.itemArray[selectedIndexPath.item]
            selectedHandle(imageName)
        }
        self.navigationController?.popViewController(animated: true)
    }
}

extension ACDModifyAvatarViewController: UICollectionViewDataSource, UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.itemArray.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDAvatarCollectionCell {
            cell.iconImageView.image = UIImage(named: self.itemArray[indexPath.item])
            cell.isSelect = self.selectedIndexPath?.item == indexPath.item
        }
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        self.selectedIndexPath = indexPath
        collectionView.reloadData()
    }
}
